### ISSUE TITLE
New-AzureRmRoleDefinition : InvalidApiVersionParameter

### ISSUE THREAD

**Issue by blackSphere-xx** [Azure/azure-powershell#3079]
### Cmdlet(s)

New-AzureRmRoleDefinition
### PowerShell Version

4.0
### Module Version

Manifest   3.2.0      AzureRM.Resources
### OS Version

 6.3.9600.17400
### Description

Running New-AzureRmRoleDefinition -InputFile "pathToJsonFile" returns InvalidApiVersionParameter
### Debug Output

Absolute Uri:
https://management.azure.com//subscriptions/88606f41-ea08-4554-8cda-99648bc55e5a/[sanitized]/providers/Microsoft.Authoriza
tion/roleDefinitions/88606f41-ea08-4554-8cda-99648bc55e5a?api-version=2015-07-01

Headers:

Body:
{
  "name": "88606f41-ea08-4554-8cda-99648bc55e5a",
  "properties": {
    "roleName": "Dev Deployment Contributor",
    "description": "Can contribute to orhub development resourcegroups, but not production groups",
    "type": "CustomRole",
    "permissions": [
      {
        "actions": [
          "Microsoft.Authorization/_/read",
          "Microsoft.Insights/alertRules/_",
          "Microsoft.Insights/diagnosticSettings/_",
          "Microsoft.Insights/components/_",
          "Microsoft.DocumentDb/databaseAccounts/_",
          "Microsoft.ResourceHealth/availabilityStatuses/read",
          "Microsoft.Resources/deployments/_",
          "Microsoft.Resources/subscriptions/resourceGroups/read",
          "Microsoft.Storage/storageAccounts/_",
          "Microsoft.Web/certificates/_",
          "Microsoft.Web/listSitesAssignedToHostName/read",
          "Microsoft.Web/serverFarms/join/action",
          "Microsoft.Web/serverFarms/read",
          "Microsoft.Web/sites/_",
          "Microsoft.Web/serverFarms/_"
        ],
        "notActions": []
      }
    ],
    "assignableScopes": [
      "/subscriptions/[sanitized]/[sanitized]
    ]
  }
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
BadRequest

Headers:
Pragma                        : no-cache
x-ms-failure-cause            : gateway
x-ms-request-id               : 87dbbd86-90a3-4adf-86b6-6dcf477cc03e
x-ms-correlation-request-id   : 87dbbd86-90a3-4adf-86b6-6dcf477cc03e
x-ms-routing-request-id       : WESTUS:20161013T201324Z:87dbbd86-90a3-4adf-86b6-6dcf477cc03e
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Cache-Control                 : no-cache
Date                          : Thu, 13 Oct 2016 20:13:24 GMT

Body:
{
  "error": {
    "code": "InvalidApiVersionParameter",
    "message": "The api-version '2015-07-01' is invalid. The supported versions are
'2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,
2014-02-26,2014-04'."
  }
}

New-AzureRmRoleDefinition : InvalidApiVersionParameter: The api-version '2015-07-01' is invalid. The supported
versions are '2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-0
1,2013-03-01,2014-02-26,2014-04'.
At line:1 char:1
- New-AzureRmRoleDefinition -InputFile .\Desktop\dev_deployment_role.json
- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : CloseError: (:) [New-AzureRmRoleDefinition], CloudException
  - FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.NewAzureRoleDefinitionCommand

DEBUG: 8:13:24 PM - NewAzureRoleDefinitionCommand end processing.
DEBUG: 8:13:24 PM - NewAzureRoleDefinitionCommand end processing.
### Script/Steps for Reproduction

Create jsonfile with contents:
{  
    "Name":  "Dev Deployment Contributor",  
    "IsCustom":  true,
    "Description":  "Can contribute to development resourcegroups, but not production groups",
    "Actions":  [
                    "Microsoft.Authorization/_/read",
                    "Microsoft.Insights/alertRules/_",
                    "Microsoft.Insights/diagnosticSettings/_",
                    "Microsoft.Insights/components/_",
                    "Microsoft.DocumentDb/databaseAccounts/_",
                    "Microsoft.ResourceHealth/availabilityStatuses/read",
                    "Microsoft.Resources/deployments/_",
                    "Microsoft.Resources/subscriptions/resourceGroups/read",
                    "Microsoft.Storage/storageAccounts/_", 
                    "Microsoft.Web/certificates/_",
                    "Microsoft.Web/listSitesAssignedToHostName/read",
                    "Microsoft.Web/serverFarms/join/action",
                    "Microsoft.Web/serverFarms/read",
                    "Microsoft.Web/sites/_",
                    "Microsoft.Web/serverFarms/_"
                ],
    "NotActions":  [

```
               ],
"AssignableScopes":  [
                         "/subscriptions/[sanitized]/[resourceGroup]"
                     ]
```

}

Run Login-AzureRMAccount (with account owner/AD Admin in my case)
Run New-AzureRmRoleDefinition -InputFile [path to json]

Output:
New-AzureRmRoleDefinition : InvalidApiVersionParameter: The api-version '2015-07-01' is invalid. The supported
versions are '2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-0
1,2013-03-01,2014-02-26,2014-04'.
At line:1 char:1
- New-AzureRmRoleDefinition -InputFile .\Desktop\Dev_Deployment_Contrib ...
- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : CloseError: (:) [New-AzureRmRoleDefinition], CloudException
  - FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.NewAzureRoleDefinitionCommand

**Comment by cormacpayne** [comment]
@ravbhatnagar Hey Gaurav, would you mind taking a look at this issue?

**Comment by bmoore-msft** [comment]
Let us know if there's a workaround - seems to be a blocking issue.

**Comment by markcowl** [comment]
@shuagarw Can you take a look?

@blackSphere-xx @bmoore-msft It would be useful to have debug logs for this, which would show the traffic with the server/  Can you execute this with ```$DebugPreference="Continue"``` and include the logs?

**Comment by blackSphere-xx** [comment]
Was the dump above insufficient? If not I'll have to find a few moments to rerun and verify that there's no information that should not be publically visible in the logs.

-----Original Message-----

**Comment by blackSphere-xx** [comment]
Full dump below. GUIDS, tokens and usernames replaced with  '##########'
=============

New-AzureRmRoleDefinition -InputFile .\Desktop\Dev_Deployment_Contributor.json
DEBUG: 6:59:54 AM - NewAzureRoleDefinitionCommand begin processing with ParameterSet 'InputFileParameterSet'.
DEBUG: 6:59:54 AM - using account id '##########'...
DEBUG: [Common.Authentication]: Authenticating using Account: '##########', environment: 'AzureCloud', tenant:
'##########',
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54:  - TokenCache: Deserialized 4 items to token cache.
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - AcquireTokenHandlerBase: === Token Acquisition
started:
 Authority: https://login.microsoftonline.com/##########/
 Resource: https://graph.windows.net/
 ClientId:'##########',
 CacheType: Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache (4 items)
 Authentication Target: User

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: Looking up cache for a token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: A Multi Resource Refresh Token for a
different resource was found which can be used
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: Multi resource refresh token for
resource 'https://management.core.windows.net/' will be used to acquire token for 'https://graph.windows.net/'
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: A matching item (access token or refresh
 token or both) was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - <RefreshAccessTokenAsync>d__11: Refreshing access
token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: Storing token in the cache...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - TokenCache: An item was stored in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: b7179b23-ad70-4a89-83ac-53510805def1 - AcquireTokenHandlerBase: === Token Acquisition
finished successfully. An access token was retuned:
 Access Token Hash: ##########
 Refresh Token Hash: ##########
 Expiration Time: 10/19/2016 14:59:53 +00:00
 User Hash: ##########

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54:  - TokenCache: Serializing token cache with 5 items.
DEBUG: [Common.Authentication]: Authenticating using Account: '##########', environment: 'AzureCloud', tenant:
'##########'
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54:  - TokenCache: Deserialized 5 items to token cache.
DEBUG: [Common.Authentication]: Authenticating using configuration values: Domain:
'##########', Endpoint: 'https://login.microsoftonline.com/', ClientId:
'##########', ClientRedirect: 'urn:ietf:wg:oauth:2.0:oob', ResourceClientUri:
'https://management.core.windows.net/', ValidateAuthrity: 'True'
DEBUG: [Common.Authentication]: Acquiring token using context with Authority
'https://login.microsoftonline.com/##########/', CorrelationId:
'00000000-0000-0000-0000-000000000000', ValidateAuthority: 'True'
DEBUG: [Common.Authentication]: Acquiring token using AdalConfiguration with Domain:
'##########', AdEndpoint: 'https://login.microsoftonline.com/', ClientId:
'##########', ClientRedirectUri: urn:ietf:wg:oauth:2.0:oob
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - AcquireTokenHandlerBase: === Token Acquisition
started:
 Authority: https://login.microsoftonline.com/106d8ba9-3f64-497a-b232-7fff1d308c4d/
 Resource: https://management.core.windows.net/
 ClientId: 1950a258-227b-4e31-a9cf-717495945fc2
 CacheType: Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache (5 items)
 Authentication Target: User

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - TokenCache: Looking up cache for a token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - TokenCache: An item matching the requested resource
was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - TokenCache: 58.8281329516667 minutes left until
token in cache expires
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - TokenCache: A matching item (access token or refresh
 token or both) was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54: 51c5fe16-8da6-4380-b1ca-570f33e19a42 - AcquireTokenHandlerBase: === Token Acquisition
finished successfully. An access token was retuned:
 Access Token Hash: ##########
 Refresh Token Hash: ##########
 Expiration Time: 10/19/2016 14:58:44 +00:00
 User Hash: ##########

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/19/2016 13:59:54:  - TokenCache: Serializing token cache with 5 items.
DEBUG: [Common.Authentication]: Received token with LoginType 'LiveId', Tenant: '##########',
 UserId: '##########'
DEBUG: [Common.Authentication]: Renewing Token with Type: 'Bearer', Expiry: '10/19/2016 14:58:44 +00:00',
MultipleResource? 'True', Tenant: '##########', UserId: '##########'
DEBUG: [Common.Authentication]: User info for token DisplayId: '##########', Name: ##########,
IdProvider: 'live.com', Uid: '##########'
DEBUG: [Common.Authentication]: Checking token expiration, token expires '10/19/2016 14:58:44 +00:00' Comparing to
'10/19/2016 13:59:54 +00:00' With threshold '00:05:00', calculated time until token expiry: '00:58:49.6654751'
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
PUT

Absolute Uri:
https://management.azure.com//subscriptions/##########/appdev/providers/Microsoft.Authoriza
tion/roleDefinitions/fab475a0-90e1-43f9-b754-5e639c800e8f?api-version=2015-07-01

Headers:

Body:
{
  "name": "fab475a0-90e1-43f9-b754-5e639c800e8f",
  "properties": {
    "roleName": "Dev Deployment Contributor",
    "description": "Can contribute to development resourcegroups, but not production groups",
    "type": "CustomRole",
    "permissions": [
      {
        "actions": [
          "Microsoft.Authorization/*/read",
          "Microsoft.Insights/alertRules/*",
          "Microsoft.Insights/diagnosticSettings/*",
          "Microsoft.Insights/components/*",
          "Microsoft.DocumentDb/databaseAccounts/*",
          "Microsoft.ResourceHealth/availabilityStatuses/read",
          "Microsoft.Resources/deployments/*",
          "Microsoft.Resources/subscriptions/resourceGroups/read",
          "Microsoft.Storage/storageAccounts/*",
          "Microsoft.Web/certificates/*",
          "Microsoft.Web/listSitesAssignedToHostName/read",
          "Microsoft.Web/serverFarms/join/action",
          "Microsoft.Web/serverFarms/read",
          "Microsoft.Web/sites/*",
          "Microsoft.Web/serverFarms/*"
        ],
        "notActions": []
      }
    ],
    "assignableScopes": [
      "/subscriptions/##########/appdev",
    ]
  }
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
BadRequest

Headers:
Pragma                        : no-cache
x-ms-failure-cause            : gateway
x-ms-request-id               : f0d2d892-f2fa-41e5-a444-4118a4d0fb1f
x-ms-correlation-request-id   : f0d2d892-f2fa-41e5-a444-4118a4d0fb1f
x-ms-routing-request-id       : EASTUS:20161019T135953Z:f0d2d892-f2fa-41e5-a444-4118a4d0fb1f
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Cache-Control                 : no-cache
Date                          : Wed, 19 Oct 2016 13:59:53 GMT

Body:
{
  "error": {
    "code": "InvalidApiVersionParameter",
    "message": "The api-version '2015-07-01' is invalid. The supported versions are
'2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,
2014-02-26,2014-04'."
  }
}

New-AzureRmRoleDefinition : InvalidApiVersionParameter: The api-version '2015-07-01' is invalid. The supported
versions are '2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-0
1,2013-03-01,2014-02-26,2014-04'.
At line:1 char:1
+ New-AzureRmRoleDefinition -InputFile .\Desktop\Dev_Deployment_Contrib ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [New-AzureRmRoleDefinition], CloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.NewAzureRoleDefinitionCommand

DEBUG: 6:59:55 AM - NewAzureRoleDefinitionCommand end processing.
DEBUG: 6:59:55 AM - NewAzureRoleDefinitionCommand end processing.

**Comment by blackSphere-xx** [comment]
@shuagarw @markcowl was the debug output above sufficient? Do you need anything else from me? I'm pretty desperate for progress here, so anything I can do to help out, please let me know. 

Any info you can give me too, even a reasonable workaround would be really, really helpful.

**Comment by shuagarw** [comment]
@RBACAsk for FYI

@blackSphere-xx  My apologies for not looking into it before as I was on vacation at the time but then never looked at it afterwards. Is this still an issue?
The error here is misleading, usually what I have seen is that if the input roleDefinition is not correct, then it throws this generic misleading error. If its still happening, share with me the roleDefinition json object and will take it from there.

**Comment by begoldsm** [comment]
I am getting the same issue for New-AzureRMRoleAssignment and Get-AzureRMRoleAssignment when I am trying to assign the "Contributor" role.

Here is the full error output:
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
GET

Absolute Uri:
https://management.azure.com//subscriptions/04319d6d-4a66-4701-bb2f-e7dbbd9ae4db/resourceGroups/begoldsmadlsrg01/begold
smadls01/providers/Microsoft.Authorization/roleDefinitions?$filter=roleName eq 'Contributor'&api-version=2015-07-01

Headers:

Body:


DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
BadRequest

Headers:
Pragma                        : no-cache
x-ms-failure-cause            : gateway
x-ms-request-id               : 3c34672c-cf00-4af6-983f-585129912dcb
x-ms-correlation-request-id   : 3c34672c-cf00-4af6-983f-585129912dcb
x-ms-routing-request-id       : WESTUS2:20170426T004755Z:3c34672c-cf00-4af6-983f-585129912dcb
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Cache-Control                 : no-cache
Date                          : Wed, 26 Apr 2017 00:47:55 GMT

Body:
{
  "error": {
    "code": "InvalidApiVersionParameter",
    "message": "The api-version '2015-07-01' is invalid. The supported versions are
'2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,
2013-03-01,2014-02-26,2014-04'."
  }
}

**Comment by RBACAsk** [comment]
The highlighted part in the Uri could be the issue.

https://management.azure.com//subscriptions/04319d6d-4a66-4701-bb2f-e7dbbd9ae4db/resourceGroups/begoldsmadlsrg01/begold<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fmanagement.azure.com%2F%2Fsubscriptions%2F04319d6d-4a66-4701-bb2f-e7dbbd9ae4db%2FresourceGroups%2Fbegoldsmadlsrg01%2Fbegold&data=02%7C01%7CSurabhi.Pandey%40microsoft.com%7C01d39dd87b174f37336108d48c3e8dae%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636287647690559084&sdata=B51kRET3VF0N1at0fyKs9GN0oBT3EkP7dHvIGzxM1DE%3D&reserved=0>smadls01/providers/Microsoft.Authorization/roleDefinitions?$filter=roleName eq 'Contributor'&api-version=2015-07-01

Thx
Surabhi

**Comment by nahalf** [comment]
Seeing this issue again with New-AzureRmRoleAssignment:
![image](https://cloud.githubusercontent.com/assets/18647532/26258960/c1c1bd5a-3c7b-11e7-84af-b1f3c5103974.png)
This is in Azure Automation.
In PowerShell locally, I am getting another error:
![image](https://cloud.githubusercontent.com/assets/18647532/26258976/d62f54b4-3c7b-11e7-9bef-18b2dc4762a5.png)

Seems like an API issue because I am using the latest version of AzureRM, and these cmdlets worked fine yesterday. I didn't change any code in the script.

**Comment by nahalf** [comment]
Passing in JSON as input to the runbook like this:

 "{\"SubscriptionName\": \"s00161csubnp000001\", \"roles\": [ { \"AccountName\": \"jpeacock@starbucks.com\", \"AccountType\": \"U\", \"RoleDefinitionName\": \"SBUX Contributor\", \"ResourceGroupName\": \"\" } ]}"

Receiving an error on these two lines:

$assignment = get-AzureRmRoleAssignment  -Scope "/subscriptions/$subscriptionID/resourceGroups/$($role.ResourceGroupName)" -RoleDefinitionName $role.RoleDefinitionName -ErrorAction Continue

*Incorrect API for the Get

And same error but for New-AzureRmRoleAssignment:

New-AzureRmRoleAssignment -ObjectId $id -Scope "/subscriptions/$subscriptionID/resourceGroups/$($role.ResourceGroupName)" -RoleDefinitionName $role.RoleDefinitionName -ErrorAction Continue

**Comment by RBACAsk** [comment]
+Hari – can you take a look?

**Comment by RBACAsk** [comment]
Sure, I shall look into the issue and get back.

From: Stefan Miltchev
Sent: Friday, May 19, 2017 12:52 PM
To: Azure/azure-powershell <reply@reply.github.com>; Azure/azure-powershell <azure-powershell@noreply.github.com>; Haridarshan Hosakere Sudhakar <hahosake@microsoft.com>
Cc: RBAC Ask <RBACAsk@microsoft.com>; Mention <mention@noreply.github.com>
Subject: RE: [Azure/azure-powershell] New-AzureRmRoleDefinition : InvalidApiVersionParameter (#3079)

+Hari – can you take a look?

**Comment by RBACAsk** [comment]
The screen-shot does not have the full Scope parameter visible. Can you please share the exact parameters passed to the cmdlet and also the precise time-range when the issue was encountered?

Thx
Surabhi

From: Haridarshan Hosakere Sudhakar
Sent: Sunday, May 21, 2017 10:33 PM
To: Stefan Miltchev <Stefan.Miltchev@microsoft.com>; Azure/azure-powershell <reply@reply.github.com>; Azure/azure-powershell <azure-powershell@noreply.github.com>
Cc: RBAC Ask <RBACAsk@microsoft.com>; Mention <mention@noreply.github.com>
Subject: RE: [Azure/azure-powershell] New-AzureRmRoleDefinition : InvalidApiVersionParameter (#3079)

Sure, I shall look into the issue and get back.

From: Stefan Miltchev
Sent: Friday, May 19, 2017 12:52 PM
To: Azure/azure-powershell <reply@reply.github.com>; Azure/azure-powershell <azure-powershell@noreply.github.com<mailto:azure-powershell@noreply.github.com>>; Haridarshan Hosakere Sudhakar <hahosake@microsoft.com<mailto:hahosake@microsoft.com>>
Cc: RBAC Ask <RBACAsk@microsoft.com<mailto:RBACAsk@microsoft.com>>; Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: RE: [Azure/azure-powershell] New-AzureRmRoleDefinition : InvalidApiVersionParameter (#3079)

+Hari – can you take a look?

**Comment by nahalf** [comment]
My apologies, I couldn’t show the full view from Azure Automation output.

This is what is being attempted:

New-AzureRmRoleAssignment -ObjectId $id -Scope "/subscriptions/$subscriptionID/resourceGroups/$($role.ResourceGroupName)" -RoleDefinitionName $role.RoleDefinitionName -ErrorAction Continue

The $role object has the following properties and this what was passed in that instande:
"roles": [ { "AccountName": "jpeacock@starbucks.com", "AccountType": "U", "RoleDefinitionName": "SBUX Contributor", "ResourceGroupName": "" } ]}’

This initially took place Friday, May 19th, around 9:15 am PST.


Natasha Halfin
Consultant
Secure Infrastructure | Azure Infrastructure
Microsoft Services
+1 (210) 421-7169 | nahalf@microsoft.com<mailto:nahalf@microsoft.com>

[Microsoft Logo]

**Comment by RBACAsk** [comment]
I noticed that the ResourceGroupName is empty. That might be the problem. Can you try with scope set as "/subscriptions/$subscriptionID”?

Thx
Surabhi

**Comment by nahalf** [comment]
So the runbook has conditional statements to allow passing ResourceGroup as empty if the user wishes to only do subscription level scope. The logic in the conditional statement turned out to be incorrect (causing it to attempt to assign an RG level assignment when it should have been a sub level one); after fixing this, the runbook is working now.

However, the error message received in that scenario is not particularly relevant or descriptive to what the actual issue was. My feedback would be if a user attempts a scope that does not exist to return an error along the lines of “Scope <scopePassed> does not exist”.


Thanks for your time and help.

Natasha Halfin
Consultant
Secure Infrastructure | Azure Infrastructure
Microsoft Services
+1 (210) 421-7169 | nahalf@microsoft.com<mailto:nahalf@microsoft.com>

[Microsoft Logo]

**Comment by RBACAsk** [comment]
+ Ming to look into this.


Thx
Surabhi

**Comment by darshanhs90** [comment]
Closed as this was resolved with V 4.3.1
